storage/redis/peer_store.go

```diff
@@ -0,0 +1,787 @@
+	} else if len(peer.IP.IP) == net.IPv6len { // implies toReturn.IP.To4() == nil
+		peer.IP.AddressFamily = bittorrent.IPv6
+	} else {
+		panic("IP is neither v4 nor v6")
+	}
+
+	return peer
+}
+
+// RedisBackend represents a Memcache result backend
+type RedisBackend struct {
+	host     string
+	password string
+	db       int
+	pool     *redis.Pool
+	// If set, path to a socket file overrides hostname
+	socketPath string
+	redsync    *redsync.Redsync
+	common.RedisConnector
+}
+
+type peerStore struct {
+	cfg Config
+	rb  *RedisBackend
+
+	closed chan struct{}
+	wg     sync.WaitGroup
+}
+
+var _ storage.PeerStore = &peerStore{}
+
+// populateProm aggregates metrics over all shards and then posts them to
+// prometheus.
+func (ps *peerStore) populateProm() {
+	var numInfohashes, numSeeders, numLeechers uint64
+
+	shards := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	for _, shard := range shards {
+		infohashes_list, err := conn.Do("HKEYS", shard) // key
+		if err != nil {
+			return
+		}
+		infohashes := infohashes_list.([]interface{})
+
+		InfohashLPrefix := shard + "_L_"
+		InfohashSPrefix := shard + "_S_"
+		InfohashPrefixLen := len(InfohashLPrefix)
+		InfohashesMap := make(map[string]bool)
+
+		for _, ih := range infohashes {
+			ih_str := string(ih.([]byte))
+			ih_str_infohash := ih_str[InfohashPrefixLen:]
+			if strings.HasPrefix(ih_str, InfohashLPrefix) {
+				numLeechers++
+				InfohashesMap[ih_str_infohash] = true
+			} else if strings.HasPrefix(ih_str, InfohashSPrefix) {
+				numSeeders++
+				InfohashesMap[ih_str_infohash] = true
+			} else {
+				log.Error("storage: invalid Redis state", log.Fields{
+					"Hkey":   shard,
+					"Hfield": ih_str,
+				})
+			}
+		}
+
+		numInfohashes += uint64(len(InfohashesMap))
+	}
+
+	storage.PromInfohashesCount.Set(float64(numInfohashes))
+	storage.PromSeedersCount.Set(float64(numSeeders))
+	storage.PromLeechersCount.Set(float64(numLeechers))
+
+	log.Debug("storage: populateProm() aggregates metrics over all shards", log.Fields{
+		"numInfohashes": float64(numInfohashes),
+		"numSeeders":    float64(numSeeders),
+		"numLeechers":   float64(numLeechers),
+	})
+}
+
+// recordGCDuration records the duration of a GC sweep.
+func recordGCDuration(duration time.Duration) {
+	log.Debug("storage: recordGCDuration", log.Fields{"timeTaken(ms)": float64(duration.Nanoseconds()) / float64(time.Millisecond)})
+	storage.PromGCDurationMilliseconds.Observe(float64(duration.Nanoseconds()) / float64(time.Millisecond))
+}
+
+func (ps *peerStore) getClock() int64 {
+	return timecache.NowUnixNano()
+}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	IPver := p.IP.AddressFamily.String()
+	log.Debug("storage: PutSeeder", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), IPver, p.Port),
```

Comment:
This will call `fmt.Sprintf` every time, regardless of whether debug logging is enabled or not - we'll have to figure out something else here, but that's also not a priority rn.

Reply:
@mrd0ll4r `log.Debug` will works only when debug logging is enabled.
```
A customizable, multi-protocol BitTorrent Tracker

Usage:
  chihaya [flags]

Flags:
      --config string       location of configuration file (default "/etc/chihaya.yaml")
      --cpuprofile string   location to save a CPU profile
      --debug               enable debug logging
  -h, --help                help for chihaya
      --json                enable json logging
      --nocolors            disable log coloring
```